### PR TITLE
[Wav2Vec2] Fix padding mask for new architectures

### LINKED ISSUE #3227: [Wav2Vec2] Possible bug with batched input
## 🐛 Bug

When using the Wav2Vec2 model with batched input (meaning that both `source` and `padding_mask` are provided), the input is different for padded input than it would be for the same non-padded input. At the moment this accounts for all models:

- Wav2Vec 2.0 Base
- Wav2Vec 2.0 Large
- Wav2Vec 2.0 Large (LV-60)*
- Wav2Vec 2.0 Large (LV-60) + Self Training *

IMO, the Base & Large models cannot really be fixed because the convolutional layers apply group norm over the time dimension meaning that adding 0's as padding tokens necessarily changes the result. See the following code snippet.

On the other hand the "new architectures" Large Lv-60 + Large Lv-60 + Self Training can be fixed as shown in #3228.

### To Reproduce

Please run the following code snippet to reproduce the error:

```python
import fairseq
import torch

# get model
wav2vec_path = "data/wav2vec_small_960h.pt"
model, cfg, task = fairseq.checkpoint_utils.load_model_ensemble_and_task(
    [wav2vec_path], arg_overrides={"data": "./data"}
)
model = model[0]
model.eval()

# create single input
input_wav_0 = torch.randn((1, 2000))
input_wav_1 = torch.randn((1, 3000))

# create batched input
batch_input_wav = torch.zeros((2, 3000))
batch_input_wav[0, :input_wav_0.shape[-1]] = input_wav_0
batch_input_wav[1, :input_wav_1.shape[-1]] = input_wav_1

# create padding mask
padding_mask = torch.zeros((2, 3000), dtype=torch.bool)
padding_mask[0, input_wav_0.shape[-1]:] = True

# run batch & single
output = model(source=input_wav_0, padding_mask=None)["encoder_out"]
batch_output = model(source=batch_input_wav, padding_mask=padding_mask)["encoder_out"]

# is equal?
print("Is batched forward and simple forward equal?", torch.allclose(output[:,0], batch_output[:output.shape[0], 0], atol=1e-1))
```

**Note**: It is assumed that both `https://dl.fbaipublicfiles.com/fairseq/wav2vec/dict.ltr.txt` and `https://dl.fbaipublicfiles.com/fairseq/wav2vec/wav2vec_small_960h.pt` were downloaded and stored in the folder `data`.

Also see [this](https://colab.research.google.com/drive/1rvP8N5WmQ4zReoH4yOAXb-ZjBSfsUZ09?usp=sharing) notebook that can be simply run to reproduce the error.

### Expected behavior

Both the padded and identical non-padded input should be equal in my opinion -> it seems like this is however not possible due to the layer norm / group norm layers in the model.

### Environment

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.7.0+cu101
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): !pip install git+https://github.com/pytorch/fairseq
 - Build command you used (if compiling from source): just pip
 - Python version: 3.6.9
 - CUDA/cuDNN version: No CUDA
 - GPU models and configuration: NO GPU

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #3227

All models that do **not** make use of group norm, such as 
- Wav2Vec 2.0 Large (LV-60)*
- Wav2Vec 2.0 Large (LV-60) + Self Training *

do need this fix IMO to able to correctly run batches through the model. Before this PR, the 
following code snippet failed:

```python
import fairseq
import torch

# get model
wav2vec_path = "data/wav2vec2_vox_960h_new.pt"
model, cfg, task = fairseq.checkpoint_utils.load_model_ensemble_and_task(
    [wav2vec_path], arg_overrides={"data": "./data"}
)
model = model[0]
model.eval()

# create single input
input_wav_0 = torch.randn((1, 2000))
input_wav_1 = torch.randn((1, 3000))

# create batched input
batch_input_wav = torch.zeros((2, 3000))
batch_input_wav[0, :input_wav_0.shape[-1]] = input_wav_0
batch_input_wav[1, :input_wav_1.shape[-1]] = input_wav_1

# create padding mask
padding_mask = torch.zeros((2, 3000), dtype=torch.bool)
padding_mask[0, input_wav_0.shape[-1]:] = True

# run batch & single
output = model(source=input_wav_0, padding_mask=None)["encoder_out"]
batch_output = model(source=batch_input_wav, padding_mask=padding_mask)["encoder_out"]

# is equal?
print("Is batched forward and simple forward equal?", torch.allclose(output[:,0], batch_output[:output.shape[0], 0], atol=1e-3))
```
Note: It is assumed that both https://dl.fbaipublicfiles.com/fairseq/wav2vec/dict.ltr.txt and https://dl.fbaipublicfiles.com/fairseq/wav2vec/wav2vec2_vox_960h_new.pt were downloaded and stored in the folder data.

Also, see [this](https://colab.research.google.com/drive/1ASZ4lVZbKkj-dvRHDl1lo0mCcsaOERlG?usp=sharing) notebook for reproducibility.

This PR should fix the behavior and make the above code snippet / notebook run succesfully.

## PR review    

Gently pinging @alexeib for Wav2Vec2

Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by alexeib — APPROVED** [review]
thanks, looks good!
overall it probably shouldnt make much difference to transcriptions that end up being output, but still a good fix.

**Comment by jeffxtang** [comment]
Will this fix resolve the issue https://github.com/pytorch/fairseq/issues/3278? Will the code snippet there be affected by this fix? I'm not sure if fairseq is required to run the code:
```
import soundfile as sf
import torch
from transformers import Wav2Vec2ForMaskedLM, Wav2Vec2Tokenizer

tokenizer = Wav2Vec2Tokenizer.from_pretrained("facebook/wav2vec2-base-960h")
model = Wav2Vec2ForMaskedLM.from_pretrained("facebook/wav2vec2-base-960h")
model.eval()
audio_input, _ = sf.read("ionlywishtobealone.wav")
input_values = tokenizer(audio_input, return_tensors="pt").input_values
logits = model(input_values).logits
predicted_ids = torch.argmax(logits, dim=-1)
transcription = tokenizer.batch_decode(predicted_ids)[0]
print(transcription)
print(input_values.shape)

# pad the 1.42s audio to about 3,5,7,9s:
for n in [3,5,7,9]:
  input_values = tokenizer(audio_input, return_tensors="pt", padding="max_length", max_length=n*10000).input_values
  logits = model(input_values).logits
  predicted_ids = torch.argmax(logits, dim=-1)
  transcription = tokenizer.batch_decode(predicted_ids)[0]
  print(input_values.shape)
  print("padding to {}s: {}\n".format(n, transcription))
```

**Comment by jeffxtang** [comment]
I just uninstalled fairseq and the code above still runs. Any ideas how I can apply the fix to the issue above, if the fix can resolve it?

**Comment by patrickvonplaten** [comment]
As far as I understand it, this fix only worked for the `lv60` models and not for the base models. The base models cannot really give the same results for padding vs. non-padding I think
